Port alert component to new approach

euclio opened the pull request:
Part of #101.

I'm getting GTK assertions when opening the alert button with the following test code, but I'm not sure what's going wrong. This should probably be fixed before merging.

```
(alert-usage:6220): Gtk-CRITICAL **: 22:03:46.191: gtk_window_get_focus: assertion 'GTK_IS_WINDOW (window)' failed

(alert-usage:6220): Gtk-CRITICAL **: 22:03:46.191: gtk_window_get_focus: assertion 'GTK_IS_WINDOW (window)' failed

(alert-usage:6220): Gtk-CRITICAL **: 22:03:46.191: gtk_window_get_focus: assertion 'GTK_IS_WINDOW (window)' failed

(alert-usage:6220): Gtk-CRITICAL **: 22:03:46.191: gtk_window_get_focus: assertion 'GTK_IS_WINDOW (window)' failed
```

```rust
use relm4::{
    gtk::{self, prelude::*},
    send, ComponentController, ComponentParts, Controller, RelmApp, Sender, SimpleComponent,
};
use relm4_components::alert::{AlertModel, AlertMsg, AlertSettings};

struct AppModel {
    alert_dialog: Controller<AlertModel>,
}

enum AppMsg {
    Show,
}

#[relm4::component]
impl SimpleComponent for AppModel {
    type Widgets = AppWidgets;
    type InitParams = ();
    type Input = AppMsg;
    type Output = ();

    view! {
        gtk::ApplicationWindow {
            gtk::Box {
                gtk::Button {
                    set_label: "Show alert",
                    connect_clicked(input) => move |_| {
                        send!(input, AppMsg::Show);
                    }
                }
            }
        }
    }

    fn init_parts(
        params: Self::InitParams,
        root: &Self::Root,
        input: &Sender<Self::Input>,
        _output: &Sender<Self::Output>,
    ) -> ComponentParts<Self> {
        let alert_dialog = AlertModel::init()
            .attach_to(root)
            .transient_for(root)
            .launch(AlertSettings {
                text: "Warning!",
                secondary_text: None,
                is_modal: true,
                destructive_accept: false,
                confirm_label: "Ok",
                cancel_label: "Cancel",
                option_label: None,
            })
            .connect_receiver(move |sender, msg| println!("{:?}", msg));

        let model = AppModel { alert_dialog };

        let widgets = view_output!();

        ComponentParts { model, widgets }
    }

    fn update(
        &mut self,
        msg: Self::Input,
        _input: &Sender<Self::Input>,
        _output: &Sender<Self::Output>,
    ) {
        match msg {
            AppMsg::Show => {
                send!(self.alert_dialog.sender(), AlertMsg::Show);
            }
        }
    }
}

fn main() {
    let app: RelmApp<AppModel> = RelmApp::new("relm4.test.alert");
    app.run(());
}

```

There seems to be some tension between messages intended to be sent from outside the component and messages for internal use only (see `AlertMsg::Response`). Is this what `Command` should be used for? `#[doc(hidden)]` doesn't seem ideal here, maybe they should be different types?

cofee-on-the-desk commented:
I've been using new-approach components for a while, here are my thoughts:

- You can get rid of the warning if you remove line with `.attach_to(root)`. It's meant to set the widget as a child for some container. You usually don't do this with windows or dialogs, use `transient_for` instead.

- You are right, `#[doc(hidden)]` seems like a hack. There's also a tangent issue: `ComponentController` exposes the internal implementation details with methods `widget` (root widget) and `state` (model / widgets).

Code notes:

- You can probably rename `AlertModel` to just `Alert`.

- Usage of `String` instead of `&'static str` would allow to generate text dynamically at runtime.

euclio commented:
Thanks for the review. I rebased, renamed `AlertModel` to `Alert`, and changed the `&'static str`s to `String`.

I think I'll leave the `#[doc(hidden)]` for now in favor of opening an issue, since it can be fixed later.

AaronErhardt commented:
Looks good to me. Yet, the recent uptick in usage of the "next" branch brought some changes recently so a rebase would be nice. I'd also be fine with just merging this and fixing it later because there will be some API breakages here and there anyway before the next release.

euclio commented:
Rebased.

AaronErhardt commented:
Thanks!
